### PR TITLE
Add support for ephemeral agent (single-use).

### PR DESCRIPTION
**Pull request by hpidcock**
This adds support for ephemeral/single-use agents. Agents given the `--ephemeral` flag will at most run one pipeline. After running a single pipeline, the agent will taint itself (which for now just disables the agent by marking it as `no_schedule`).

The purpose for this feature for agents, is to allow potentially _privileged_ pipelines to run on an agent, isolated from other pipelines. Then after the pipeline has run, effectively throw away the agent. It is intended that these agents are provisioned by a seperate system, external to woodpecker. An external system would _dynamically_ provision agents (based on number of waiting pipelines and their labels), run those agents with `--ephemeral`, wait for them to be tainted, then destroy them.

TODO:
- [ ] Move tainting to just before running the single pipeline
- [ ] Tests
- [ ] Docs

### CODE REVIEW

File: server/grpc/server.go [hunk 1]
```diff
@@ -79,7 +81,7 @@ func (s *WoodpeckerServer) Next(c context.Context, req *proto.NextRequest) (*pro
 		return res, err
 	}
 	if pipeline == nil {
-		return res, err
+		return res, status.Error(codes.Unavailable, "no pipeline available: try again")
```
[the remark this answers is not in the document]
Comment (hpidcock) [reply server/grpc/server.go:84]: I guess not. Unavailable implies the individual gRPC call is able to be retried, so I'll drop the try again.

File: agent/runner.go [hunk 2]
```diff
@@ -193,6 +205,9 @@ func (r *Runner) Run(runnerCtx context.Context) error {
 		logger.Debug().Msg("updating pipeline status complete")
 	}
 
+	if r.ephemeral {
+		return ErrAgentTainted
+	}
```
[the remark this answers is not in the document]
Comment (hpidcock) [reply agent/runner.go:210]: I've moved this handling back out to the cmd/agent runner go routine where I had it originally. 

I needed to still have an error to differentiate between no workflow and workflow completed, so I've added ErrNoWorkflow.

File: agent/runner.go [hunk 3]
```diff
@@ -62,7 +66,15 @@ func (r *Runner) Run(runnerCtx context.Context) error {
...
-		return nil
+		return ErrNoWorkflow
+	}
+
+	// if ephemeral, taint the agent before running any workload.
+	if r.ephemeral {
+		err = r.client.TaintAgent(runnerCtx)
+		if err != nil {
+			return fmt.Errorf("tainting agent: %w", err)
+		}
```
[the remark this answers is not in the document]
Comment (hpidcock) [reply agent/runner.go:77]: It is possible to move out, but the ordering is important here.

The agent needs to be tainted after receiving the workflow, just before it runs it.

- If it is tainted _before_, then it will never receive one.
- If it is tainted _after_ a workflow, and that workflow causes the agent to restart in some way, it will never be tainted.

I'm open to alternative ways we can achieve this. It could be handled server side instead, but I was trying as much as possible to limit the scope of the changes.

**Comment by anbraten** [comment]
Did you had a look at the woodpecker autoscaler. Its an external tool which is also watching agents and disabling them  based on a specific condition. So I guess a similar tool could solve your usecase without needing to adjust the core

File: agent/runner.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hpidcock) [reply agent/runner.go:77]: This is only probable with a malicious workflow. If there was a workflow using privileged steps or was running on a local backend agent, it could potentially take over the agent before it is "tainted", and pull whatever workflows it wants until it gets some information or run some other nefarious action.

My goal with this work is to enable me to run privileged steps on my agents by making them ephemeral. This is the first step in doing that.

**Comment by hpidcock** [comment]
> Did you had a look at the woodpecker autoscaler. Its an external tool which is also watching agents and disabling them based on a specific condition. So I guess a similar tool could solve your usecase without needing to adjust the core

I have looked at this tool, it is what I want in terms of the creation/destruction of the agents (I'll need to add AWS+OpenStack support for my needs). But this PR's use case is specifically that I cannot trust an agent after it has started a workflow, and I need it to be able limit itself from running any more workflows.

**Comment by wez** [comment]
Just came here to say: I'm looking for this single-use agent feature, and here's the context for my use case:

Specifically what I'm looking for is:
1. Run a single workflow, privileged
2. Reset the machine containing the agent to a known-good snapshot
3. Repeat

I intend on running this in a local proxmox so that I can apply appropriate isolation (vlan + firewall to prevent access to other networks)

I'm open to other approaches that would enable me to run privileged workflows on my local hardware, with the same kind of isolation mentioned above, and without any state on the agent leaking between workflow runs.

**Comment by zc-devs** [comment]
> pipelines to run on an agent

> I cannot trust an agent after it has started a workflow, and I need it to be able limit itself from running any more workflows

Seems, you are talking about `local` backend.

> I'm open to other approaches...

It might be
1. [libvirt backend](https://github.com/woodpecker-ci/woodpecker/discussions/2302);
2. [kubevirt](https://github.com/kubevirt/kubevirt) backend;
3. or you can try to [run Kata Containers with Kubernetes](https://github.com/kata-containers/kata-containers/blob/main/docs/how-to/run-kata-with-k8s.md) now, maybe it would work with `docker` backend also.

**Comment by theanurin** [comment]
I'm interested in the feature too.

My use-case is same to described by @wez.

I'm migrating GitLab -> Woodpecker.
My setup described here https://aljax.us/how-to-setup-gitlab-runners-in-kvm-qemu-virtual-machines/
This feature may cover [GitLab's `cleanup_exec`](https://docs.gitlab.com/runner/executors/custom.html#cleanup)

I'm happy to help with the PR!
I start from testing this implementation in my env...

**Comment by 6543** [comment]
well please resolve conficts  :)

**Comment by 6543** [comment]
through waiting for  #3895 might be a good idea ...

**Comment by qwerty287** [comment]
superseded by #6150
